File: frontend/calmed-tics/iosApp/iosApp/TransportBarObserver.swift
import AVKit
import Shared

final class TransportBarObserverImpl: NSObject, TransportBarObserver, AVPlayerViewControllerDelegate {

    func observe(controller: AVPlayerViewController) {
        controller.delegate = self
    }

    func playerViewController(
        _ playerViewController: AVPlayerViewController,
        willTransitionToVisibilityOfTransportBar visible: Bool,
        with coordinator: AVPlayerViewControllerAnimationCoordinator
    ) {
        TransportBarObserverHolder.shared.notifyVisibilityChanged(visible: visible)
    }
}

File: frontend/calmed-tics/shared/src/iosMain/kotlin/com/calmed/calmedtics/ui/component/TransportBarObserver.kt
@file:OptIn(kotlinx.cinterop.ExperimentalForeignApi::class)

package com.calmed.calmedtics.ui.component

import platform.AVKit.AVPlayerViewController

/**
 * Implemented in Swift. The Kotlin/Native AVKit bindings don't expose
 * [AVPlayerViewControllerDelegate.willTransitionToVisibilityOfTransportBar],
 * so the Swift side attaches itself as the controller's delegate and reports
 * transport-bar visibility back through [TransportBarObserverHolder].
 */
interface TransportBarObserver {
    fun observe(controller: AVPlayerViewController)
}

object TransportBarObserverHolder {
    var observer: TransportBarObserver? = null

    var onVisibilityChanged: ((Boolean) -> Unit)? = null

    fun notifyVisibilityChanged(visible: Boolean) {
        onVisibilityChanged?.invoke(visible)
    }
}
